File: packages/python/llm_local_client/infrastructure/model_manager.py
import requests
from typing import List

class ModelManager:
    def __init__(self, base_url: str = 'http://localhost:11434'):
        self.base_url = base_url

    def list_models(self) -> List[str]:
        try:
            response = requests.get(f"{self.base_url}/api/tags", timeout=10)
            response.raise_for_status()
            data = response.json()
            return [m.get("name") for m in data.get("models", [])]
        except requests.RequestException as e:
            raise RuntimeError(f"Failed to fetch models from Ollama: {e}")

    def ensure_model(self, model_name: str):
        models = self.list_models()
        if model_name not in models and f"{model_name}:latest" not in models:
            print(f"Model {model_name} not found locally. Attempting to pull...")
            self._pull_model(model_name)

    def _pull_model(self, model_name: str):
        try:
            response = requests.post(f"{self.base_url}/api/pull", json={"name": model_name}, timeout=300)
            response.raise_for_status()
            print(f"Successfully pulled model {model_name}.")
        except requests.RequestException as e:
            raise RuntimeError(f"Failed to pull model {model_name}: {e}")

File: packages/typescript/src/infrastructure/modelManager.ts
import axios from 'axios';

export class ModelManager {
  private baseUrl: string;

  constructor(baseUrl: string = 'http://localhost:11434') {
    this.baseUrl = baseUrl;
  }

  async listModels(): Promise<string[]> {
    try {
      const response = await axios.get(`${this.baseUrl}/api/tags`);
      return response.data.models.map((m: any) => m.name);
    } catch (error) {
      throw new Error('Failed to fetch models from Ollama.');
    }
  }

  async ensureModel(modelName: string): Promise<void> {
    const models = await this.listModels();
    if (!models.includes(modelName) && !models.includes(`${modelName}:latest`)) {
      console.log(`Model ${modelName} not found locally. Attempting to pull...`);
      await this.pullModel(modelName);
    }
  }

  private async pullModel(modelName: string): Promise<void> {
    try {
      // In a real implementation, you'd handle the streaming response to show progress
      await axios.post(`${this.baseUrl}/api/pull`, { name: modelName });
      console.log(`Successfully pulled model ${modelName}.`);
    } catch (error) {
      throw new Error(`Failed to pull model ${modelName}.`);
    }
  }
}
